src/utils/memory_pool.hpp

```diff
@@ -0,0 +1,274 @@
+//========================================================================================
+// (C) (or copyright) 2022. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef UTILS_MEMORY_POOL_HPP_
+#define UTILS_MEMORY_POOL_HPP_
+
+#include <iostream>
+#include <Kokkos_Core.hpp>
+#include <string>
+#include <vector>
+#include <iostream>
+#include <memory>
+#include <stack>
+#include <map>
+#include <unordered_map>
+#include <type_traits>
+#include <math.h>
+#include <tuple>
+#include <utility>
+
+namespace parthenon { 
+
+// Object for managing a pool of Kokkos::Views that
+// have the same instantiation call signature
+template <class T, class KEY_T = unsigned long int>
+class Pool
+{
+public:
+  using base_t = T;
+  // Forward declarations of pool types
+  class weak_t;
+  class owner_t;
+
+private:
+  std::function<T()> get_resource_;
+  std::stack<weak_t> available_;
+  std::unordered_map<KEY_T, std::pair<weak_t, int>> inuse_;
+  static const KEY_T default_key_ = KEY_T();
+  KEY_T keyc_;
+
+public:
+  template <class... Ts>
+  Pool(std::function<T()> get_resource) : get_resource_(get_resource),
+                                          available_(),
+                                          inuse_(),
+                                          keyc_(default_key_) {}
+
+  weak_t Get();
+
+  void PrintStatistics()
+  {
+    std::cout << available_.size() << " unused objects." << std::endl;
+    std::cout << inuse_.size() << " used objects." << std::endl;
+  }
+
+private:
+  bool IsValid(const weak_t &in) const
+  {
+    return inuse_.count(in.key_);
+  }
+
+  void ReferenceCountedFree(const weak_t &in)
+  {
+    if (!IsValid(in))
+      return;
+    auto &pair = inuse_[in.key_];
+    --pair.second;
+    if (pair.second <= 0)
+    {
+      available_.push(pair.first);
+      inuse_.erase(in.key_);
+    }
+  }
+
+  void Free(const weak_t &in)
+  {
+    if (!IsValid(in))
+      return;
+    available_.push(inuse_[in.key_].first);
+    inuse_.erase(in.key_);
+  }
+
+  void AddCount(const weak_t &in)
+  {
+    if (!IsValid(in))
+      throw 1;
+    ++inuse_[in.key_].second;
+  }
+};
```

Comment:
I love having a memory pool. This is great to have... and maybe more broadly applicable.

If I'm understanding correctly, this view pool machinery creates a new view every time it's requested one that doesn't exist. Could we potentially instead allocate a big buffer of memory and then pass out shallow slices of one big view behind the scenes? That might (slightly) enhance performance by reducing the number of mallocs.

The trade-off of course is that the book-keeping would be pretty different. Maybe keep track of total memory allocated, total memory released, and pointer to the end of the last allocation. Then do a "defrag" every while to avoid memory leaks?

To me that book-keeping design seems more intuitive and extendible/flexible than what you have here. But what you have here also makes sense.

Reply:
Yes, the current implementation checks to see if there are any free objects in the pool. If there is one, it grabs it and hands it back out. Otherwise, it allocates a new object and hands it out. When an owning object is destroyed and the reference count of an object goes back to zero, that object is released back into the pool and can be re-used. This should result in a significant number of allocations initially, but assuming the number of communication channels isn't changing too much after remeshes and sparse allocation/deallocation, there should be minimal allocation afterward since it is just grabbing previously used buffers.

Allocating a large chunk of memory up front could also work, although I don't see how the book keeping would be substantially different. You would allocate it, split it into sub-views, and put those in the free object pool. From then on, I think the logic would look similar to what is being done currently. It would also make the code less generic, since `T` could essentially only be a `View` at that point, although that is probably not a big deal since that is the main/only use case I know of right now.

Reply:
I have added some functionality to allow for this.